core/engine/src/value/mod.rs

```diff
@@ -890,19 +890,19 @@ impl JsValue {
         debug_assert!(0 <= clamped && clamped <= Number::MAX_SAFE_INTEGER as i64);
 
         // e. Return integer.
-        Ok(clamped as u64)
+        Ok(clamped as usize)
     }
 
     /// Converts argument to an integer suitable for use as the length of an array-like object.
     ///
     /// See: <https://tc39.es/ecma262/#sec-tolength>
-    pub fn to_length(&self, context: &mut Context) -> JsResult<u64> {
+    pub fn to_length(&self, context: &mut Context) -> JsResult<usize> {
         // 1. Let len be ? ToInteger(argument).
         // 2. If len ≤ +0, return +0.
         // 3. Return min(len, 2^53 - 1).
```

Comment:
We used to have it be `usize`, but we changed it because on `32-bit` systems it does not capture the full range that a JavaScript length can be. (for arrays it's fine to put it as `usize` since it's between `0..2^32-1`, but for strings and typed arrays, the upper bound is `2^53-1` (max safe representable integer in a `f64`).

Setting it to `usize` truncates the values on 32-bit platforms giving wrong results.

Reply:
For example, one place that `to_length` is used is [`String.padEnd()`](https://github.com/boa-dev/boa/blob/e529d702ec5cc869a08be7486b82e06d23d4466a/core/engine/src/builtins/string/mod.rs#L1524) on the `maxLength` parameter, putting `2^32-1` would work on 32-bit platform anything greater will be truncated (i.e. `2^32` will be truncated to `0`), on 64-bit platform this would be fine.

Reply:
MAX_STRING_LENGTH is 3^32-1.

From the MDN:
> The language specification requires strings to have a maximum length of 253 - 1 elements, which is the upper limit for precise integers. However, a string with this length needs 16384TiB of storage, which cannot fit in any reasonable device's memory, so implementations tend to lower the threshold, which allows the string's length to be conveniently stored in a 32-bit integer.
>
> In V8 (used by Chrome and Node), the maximum length is 229 - 24 (~1GiB). On 32-bit systems, the maximum length is 228 - 16 (~512MiB).
> In Firefox, the maximum length is 230 - 2 (~2GiB). Before Firefox 65, the maximum length was 228 - 1 (~512MiB).
> In Safari, the maximum length is 231 - 1 (~4GiB).

Reply:
The issue isn't with allocation of the strings, rather that it breaks the math and checks in many places.

Reply:
That's all fine. If there's a bug with math being broken, then we need to fix it.

I would assume test262 and/or our tests would find any discrepancies, and I'd consider any issue a blocker for this PR. I'm currently running the full test262 suite on a Rasp Pi 2 equivalent. Everything works locally on my aarch64 Mac.

Reply:
If you can provide me with a valid JavaScript snippet that works on other implementations but doesn't work on this PR, I'll be more than happy to add it to tests and fix this PR.